lib/matplotlib/widgets.py

```diff
@@ -706,76 +704,58 @@ def __init__(self, ax, label, initial='',
+        self.label = ax.text(
+            -label_pad, 0.5, label, transform=ax.transAxes,
+            verticalalignment='center', horizontalalignment='right')
+        self.text_disp = self.ax.text(
+            self.DIST_FROM_LEFT, 0.5, initial, transform=self.ax.transAxes,
+            verticalalignment='center', horizontalalignment='left')
 
         self.cnt = 0
         self.change_observers = {}
         self.submit_observers = {}
 
-        # If these lines are removed, the cursor won't appear the first
-        # time the box is clicked:
-        self.ax.set_xlim(0, 1)
-        self.ax.set_ylim(0, 1)
+        ax.set(
+            xlim=(0, 1), ylim=(0, 1),  # s.t. cursor appears from first click.
+            navigate=False, facecolor=color,
+            xticks=[], yticks=[])
 
         self.cursor_index = 0
 
-        # Because this is initialized, _render_cursor
-        # can assume that cursor exists.
-        self.cursor = self.ax.vlines(0, 0, 0)
-        self.cursor.set_visible(False)
+        self.cursor = ax.vlines(0, 0, 0, visible=False,
+                                transform=mpl.transforms.IdentityTransform())
 
         self.connect_event('button_press_event', self._click)
         self.connect_event('button_release_event', self._release)
         self.connect_event('motion_notify_event', self._motion)
         self.connect_event('key_press_event', self._keypress)
         self.connect_event('resize_event', self._resize)
-        ax.set_navigate(False)
-        ax.set_facecolor(color)
-        ax.set_xticks([])
-        ax.set_yticks([])
+
         self.color = color
         self.hovercolor = hovercolor
 
-        self._lastcolor = color
-
         self.capturekeystrokes = False
 
-    def _make_text_disp(self, string):
-        return self.ax.text(self.DIST_FROM_LEFT, 0.5, string,
-                            verticalalignment='center',
-                            horizontalalignment='left',
-                            transform=self.ax.transAxes)
+    @property
+    def text(self):
+        return self.text_disp.get_text()
 
     def _rendercursor(self):
         # this is a hack to figure out where the cursor should go.
         # we draw the text up to where the cursor should go, measure
         # and save its dimensions, draw the real text, then put the cursor
         # at the saved dimensions
 
-        widthtext = self.text[:self.cursor_index]
-        no_text = False
-        if widthtext in ["", " ", "  "]:
-            no_text = widthtext == ""
-            widthtext = ","
-
-        wt_disp = self._make_text_disp(widthtext)
-
-        self.ax.figure.canvas.draw()
-        bb = wt_disp.get_window_extent()
-        inv = self.ax.transData.inverted()
-        bb = inv.transform(bb)
-        wt_disp.set_visible(False)
-        if no_text:
-            bb[1, 0] = bb[0, 0]
-        # hack done
-        self.cursor.set_visible(False)
+        text = self.text
```

Comment:
This is just used twice in the method. No benefit in assigning to a separate variable. Let's just use `self.text` below.

Reply:
No, this needs to be saved because it's just a property pointing to the text_disp contents.  Rewrote that more explicitly as it's indeed confusing.